### PR TITLE
RFC: 3D Model Support

### PR DESCRIPTION
**Pull request by AetherUnbound**
Combined from WordPress/openverse-frontend#889 and WordPress/openverse-catalog#370

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: rfcs/20220221-3d_model_support.md
Comment (sarayourfriend) [inline rfcs/20220221-3d_model_support.md:180]: Copying my remarks and @zackkrida's additional background information from the issue:

I would really really like to avoid SketchFab's viewer if possible.

1. It is not free software
2. It has much [looser/worse privacy guarantees](https://help.sketchfab.com/hc/en-us/articles/203059088-Compatibility#troubleshooting-scripts) than we've tried to achieve with Openverse so far

On the other hand, their viewer is extremely convenient and works very well.

In a brief search I found this [js-3d-model-viewer](https://github.com/cgwire/js-3d-model-viewer) library that is MIT licensed. It's not great, but I wonder if we could help improve it and add some of the features we might want (like UI buttons for controlling zoom when a mousewheel isn't available, recentering, snapping to specific axis, etc).

Currently it only support OBJ files though and has poor texture support (the latter of which is something we could probably spend some time contributing against). These might be unacceptable limitations though.

I guess it depends on how much upfront work we're willing to do for this or if it's something we'd rather just get out the door and be able to move on to video.

**From Zack:**

Regarding viewers, I chatted a bit with @antpb today who works on wp core media, but also around open source projects tackling standardizing 3D content performance, and might have some recommendations for us here smile

**Review by sarayourfriend — COMMENTED** [review]
Exciting!!!

File: rfcs/20220221-3d_model_support.md
Comment (antpb) [reply rfcs/20220221-3d_model_support.md:180]: Hey there! I have been pretty heavy involved in 3D web development over the last several years and have been particularly focused on Three.js and glTF as my go-to stack for rendering 3D in a browser. I have a plugin that was released a couple weeks ago that officially supports three.js, WebXR and glTF files in the WordPress block editor. You can try the plugin here: https://wordpress.org/plugins/three-object-viewer/

I feel pretty strongly that glTF, fbx, Usdz, and obj will be the most asked for supports. React three fiber is a library that supports gltf and fbx out of the box and I have an example of an obj loader relative to three.js that I would be willing to contribute. React Three Fiber is what I'm using for the render and it is very plug and play with Gutenberg, so implementations of 3D seem like they would be more consistent with a library such as this. Hopefully my plugin will help give inspiration around how people want to use 3D files on their sites. I have many more features like animations and interactions in the pipeline!

Separately from the plugin above, I am part of a group working toward standards around 3D content on the web. The group is called Open Metaverse Interoperability Group: https://omigroup.org and it has been very successful in building a community of leaders in the open metaverse pursuit. All code within the group adheres to the w3c requirements and is licensed MIT. The group is looking for wider support of these extensions as time goes on so Openverse seems a really great pairing! 

( More on the w3c side of OMI here: https://www.w3.org/community/metaverse-interop/  )

The problem currently that the group is solving is around a lack of standards with glTF when it comes to more rich interactions (such as audio, interactive elements, etc). Every engine or renderer requires a unique implementation to achieve these features which means the content (3D object) usually only works for one. The group is solving part of this problem through glTF extensions which allow for interoperability across many environments. Currently the team has developed an audio emitter spec that has resulted in a node package that allows for adding a single parser to support both spatial and global audio in a three.js scene. These standards are not specific to three.js which is what makes the support of the gltf-extensions interoperable across many engines. An example of an extension spec is here: https://github.com/omigroup/gltf-extensions

Also here is the plug and play three.js package that enables files with these extensions to work like magic: https://www.npmjs.com/package/three-omi

A live example of this audio emitter can be found here: https://3ov.xyz/omi-audio-demo/

The above package was implemented in my plugin and you can explore the different features at my plugin's website here: https://3ov.xyz/

Hopefully this info helps. I realized I dumped a bunch of links but I'm always happy to talk about this topic as I have given it _a lot_ of thought. :D

File: rfcs/20220221-3d_model_support.md
Comment (zackkrida) [reply rfcs/20220221-3d_model_support.md:42]: Data Definition Language, so here it means the sql file that creates the 3d models table in local postgres. It's not a true migration in the sense that it's not run by a migration tool in a managed way, it's basically a schema. 

I do think we could clarify this for contributors though!

File: rfcs/20220221-3d_model_support.md
Comment (zackkrida) [reply rfcs/20220221-3d_model_support.md:122]: Or just delete the existing image data for these providers from the catalog, and then run the new provider scripts. I'll add a note about this.

File: rfcs/20220221-3d_model_support.md
Comment (zackkrida) [reply rfcs/20220221-3d_model_support.md:174]: Yeah, these items are very generic certainties, and we'll definitely need a new, more detailed RFC for implementing the designs. I'll also add a note about that here.

File: rfcs/20220221-3d_model_support.md
Comment (antpb) [reply rfcs/20220221-3d_model_support.md:16]: I think textures are hard to anticipate the UV maps for every object. There are some common objects like maybe humanoid characters that would have very similar textures for things like hair, but overall it's on an object by object basis in my experience. Things like compression also manipulate those UV maps so textures may be inconsistent this way.

Comment (antpb) [reply rfcs/20220221-3d_model_support.md:16]: I do think textures in general are a category separate from 3D useful for many cases! I use subtle textures all the time in 3D assets but I think they have more utility than just 3D objects

File: rfcs/20220221-3d_model_support.md
Comment (antpb) [reply rfcs/20220221-3d_model_support.md:102]: Number of textures comes into play when you have multiple materials. Material count has a pretty big impact on performance and often 3D browser experiences recommend around 20 materials max without having performance issues. (Like iOS just crashes sometimes) Worth looking at is the guidelines that Mozilla Hubs recommends for their 3D scenes. They mention how textures relate to performance and why they matter with a mention of sketchfab: https://hubs.mozilla.com/docs/spoke-optimization.html

Comment (antpb) [reply rfcs/20220221-3d_model_support.md:102]: When you save files in that platform, it gives you a performance readout. These might be good just to note what folks care about when creating content.

Performance Check from Spoke

**Comment by panchovm** [comment]
To @sarayourfriend's point

> attribution recommendations/guidelines might be slightly different for 3D models. In particular I'm thinking about when you use them for a game or an animated film or something, what's the correct way to give attribution?

The attribution box shows the attribution info required instead of guiding how to place it depending on the format. To your point, we might add a CC-recommended documentation link for people using 3D models. Although I have not seen anything similar.

File: rfcs/20220221-3d_model_support.md
Comment (AetherUnbound) [reply rfcs/20220221-3d_model_support.md:42]: Oops, oversight on my part! Would adding a reference here to, 1) what a DDL is and 2) a current example, help here?

File: rfcs/20220221-3d_model_support.md
Comment (AetherUnbound) [reply rfcs/20220221-3d_model_support.md:102]: Thank you for your input @antpb! I am no 3D-model expert, so any other fields you think might be useful to have here would be welcome :smile:

File: rfcs/20220221-3d_model_support.md
Comment (AetherUnbound) [reply rfcs/20220221-3d_model_support.md:122]: Yea great call, it would be easiest to purge them from the catalog's `image` table and run the new provider as Zack describes.

File: rfcs/20220221-3d_model_support.md
Comment (sarayourfriend) [reply rfcs/20220221-3d_model_support.md:42]: I feel like it's domain knowledge or something that I should have known :sweat_smile: You do already link to the folder of existing DDLs, I just didn't know that SQL was DDL.

File: rfcs/20220221-3d_model_support.md
Comment (zackkrida) [reply rfcs/20220221-3d_model_support.md:180]: Despite the fact that Openverse's frontend is written in Vue.js, it _could still be_ entirely appropriate to use react-three-fiber regardless. Thanks for your thoughts!

Comment (sarayourfriend) [reply rfcs/20220221-3d_model_support.md:180]: I don't think we should load react for this, but it does seem like there are some Vue equivalents: https://github.com/troisjs/trois

It appears to be more or less the same thing as react-three-fiber. It would need some work to get [this issue closed](https://github.com/troisjs/trois/issues/70) before we could use it though.

Comment (zackkrida) [reply rfcs/20220221-3d_model_support.md:180]: I wouldn't be surprised if some renderers are even larger than react and react-dom, so while it's unlikely it's not completely off the table for me, but Trois looks amazing! Well maintained and it would be fun to contribute anything we find missing.

Comment (sarayourfriend) [reply rfcs/20220221-3d_model_support.md:180]: That's true. I guess either way with this work we'll need to be careful to make sure we're lazy-loading/bundle splitting the rendering libraries and tools anyway; they are indeed quite heavy.

File: rfcs/20220221-3d_model_support.md
Comment (sarayourfriend) [inline rfcs/20220221-3d_model_support.md:67]: Aside: It appears that AudioSets is actually just a json blob on the Audio row.

Why is that not a foreign key I wonder?

File: rfcs/20220221-3d_model_support.md
Comment (sarayourfriend) [inline rfcs/20220221-3d_model_support.md:65]: This is probably a good question to call out for @antpb to give feedback directly on (or Anthony, if you know any other folks who'd be willing to give input here that'd be :100: as well)

File: rfcs/20220221-3d_model_support.md
Comment (sarayourfriend) [inline rfcs/20220221-3d_model_support.md:66]: I think yes, if we can get this metadata it'd be a useful filter :slightly_smiling_face:

File: rfcs/20220221-3d_model_support.md
Comment (sarayourfriend) [inline rfcs/20220221-3d_model_support.md:116]: What do we need to do for this? Does this just mean using intersection observer or something like that to prevent them from loading?

If IntersectionObserver isn't available on a platform, should we fall back to deferring loading some other way?

Do we need to try to eagerly fetch resources under certain circumstances (maybe we start loading the 3d model after hover or focusing on a result for a certain amount of time).

We probably need to invest more time into figuring out how we're going to demonstrate that the resource is still loading. We have [this issue for adding a loading animation for the audio](https://github.com/WordPress/openverse-frontend/issues/1001) (in fact I thought that issue already existed but it must have fell through the cracks when we discussed it). Whatever we do there could inform what we do for 3d models as well.

File: rfcs/20220221-3d_model_support.md
Comment (sarayourfriend) [inline rfcs/20220221-3d_model_support.md:118]: If we could flesh out more details about graceful degradation on low power devices that would be great. As mentioned above, 3d models can be heavy assets. We should avoid crashing someones low-power phone by loading too many (or potentially _any_) of them. Unless there are ways to process the assets that could accommodate low memory or small CPU; for example, is it possible to automatically decrease the number of vertices on a model or something like that?

File: rfcs/20220221-3d_model_support.md
Comment (sarayourfriend) [inline rfcs/20220221-3d_model_support.md:152]: Can we specify what "Create a feature flag" means exactly _vis_ https://github.com/WordPress/openverse/pull/171/files#r811500414. I feel like regardless of the _how_ behind the feature flag flipping mechanism it should get put inside a composable, right? Probably a `readonly`-after-initialization Pinia store?

File: rfcs/20220221-3d_model_support.md
Comment (sarayourfriend) [inline rfcs/20220221-3d_model_support.md:156]: This link will probably stale depending on the timing with #174's implementation.

File: rfcs/20220221-3d_model_support.md
Comment (AetherUnbound) [reply rfcs/20220221-3d_model_support.md:67]: Hmm, I don't think it is? It looks like it's a separate table to me: https://github.com/WordPress/openverse-api/blob/c5daff0a090883d2232394872293796734689640/api/catalog/api/models/audio.py#L186

```
openledger> \dt audio*;
+--------+------------------+-------+--------+
| Schema | Name             | Type  | Owner  |
|--------+------------------+-------+--------|
| public | audio            | table | deploy |
| public | audiolist        | table | deploy |
| public | audiolist_audios | table | deploy |
| public | audioset         | table | deploy |
+--------+------------------+-------+--------+
SELECT 4
Time: 0.087s
openledger> \d audioset;
+---------------------+--------------------------+-------------------------------------+
| Column              | Type                     | Modifiers                           |
|---------------------+--------------------------+-------------------------------------|
| id                  | integer                  |  not null                           |
| created_on          | timestamp with time zone |  not null default CURRENT_TIMESTAMP |
| updated_on          | timestamp with time zone |  not null default CURRENT_TIMESTAMP |
| foreign_identifier  | character varying(1000)  |                                     |
| title               | character varying(2000)  |                                     |
| foreign_landing_url | character varying(1000)  |                                     |
| creator             | character varying(2000)  |                                     |
| creator_url         | character varying(2000)  |                                     |
| url                 | character varying(1000)  |                                     |
| filesize            | integer                  |                                     |
| filetype            | character varying(80)    |                                     |
| thumbnail           | character varying(1000)  |                                     |
| provider            | character varying(80)    |                                     |
+---------------------+--------------------------+-------------------------------------+
Indexes:
    "api_audioset_pkey" PRIMARY KEY, btree (id)
    "api_audioset_url_key" UNIQUE, btree (url)
    "unique_foreign_identifier_provider" UNIQUE, btree (foreign_identifier, provider)
    "api_audioset_foreign_identifier_28c1db59" btree (foreign_identifier)
    "api_audioset_foreign_identifier_28c1db59_like" btree (foreign_identifier varchar_pattern_ops)
    "api_audioset_url_200f9fb4_like" btree (url varchar_pattern_ops)
    "audioset_provider_cc061f30" btree (provider)
    "audioset_provider_cc061f30_like" btree (provider varchar_pattern_ops)

Time: 0.551s
```

I could be misinterpreting your question :slightly_smiling_face:

Comment (sarayourfriend) [reply rfcs/20220221-3d_model_support.md:67]: I was going off of the DDL in the catalog, but maybe it gets turned into a real table when transferred to the API DB?

https://github.com/WordPress/openverse-catalog/blob/0c5b43fbab8cba491817ebb647b77e17114a7353/docker/local_postgres/0007_openledger_audio_view.sql#L90-L103

https://github.com/WordPress/openverse-catalog/blob/6c172033e42a91bcd8f9bf78fd6b933a70bd88bf/archive/openledger_sql/03_create_audio_schema.sql#L33-L35

Comment (AetherUnbound) [reply rfcs/20220221-3d_model_support.md:67]: Ah yea, that's exactly what happens!

https://github.com/WordPress/openverse-api/blob/1d186bcf9d1df29d917bd79847840f82cf20e0ce/ingestion_server/ingestion_server/tasks.py#L115

I can elaborate a bit more on this point to denote the distinction (since it'll likely be a similar setup).

Comment (sarayourfriend) [reply rfcs/20220221-3d_model_support.md:67]: Is that more performant for ingestion because it eliminates a FK? It seems counter-intuitive that storing _more_ data overall would be better (especially as jsonb).

Comment (AetherUnbound) [reply rfcs/20220221-3d_model_support.md:67]: I'm almost certain that's the reason that's the reason why this is in the primary table as `jsonb` and pulled out into a view, maybe @obulat and @krysal might know more?

**Comment by zackkrida** [comment]
I thought of something I need to add to the RFC:

In the frontend 'All Results' view, should we make the media type _also_ some kind of filter? Since we will have more than 2 media types, will we allow permutations like '3d models and images' or 'images and audio' (basically excluding one of the three types)? Or will we just keep the choice 'all' vs. 'single' content type?

File: rfcs/20220221-3d_model_support.md
Comment (zackkrida) [reply rfcs/20220221-3d_model_support.md:16]: Yeah, 'texture' might be a good category to add to our images, once we have a source with those kinds of assets.

File: rfcs/20220221-3d_model_support.md
Comment (zackkrida) [reply rfcs/20220221-3d_model_support.md:157]: Added a note that this depends on the feature flag RFC.

**Comment by zackkrida** [comment]
@WordPress/openverse-developers Can we get more reviews on this? If you've already looked at it in the past, please know I made some requested updates today.

**Comment by zackkrida** [comment]
@panchovm here's some design thoughts I had. I think we should add a design section to the RFC before merging it. What do you think? If you have any recommendations or tasks I didn't think of please let me know. Once you've reviewed I can add these to the RFC.

- [ ] Design/select an icon for the 3D content type
- [ ] Design how 3D models look in the 'all results' view
  - Need a way of distinguishing 3D models from images, since we will only show model thumbnails
- [ ] Design results grid when the chosen content type is '3D models'
- [ ] Design 3D model single result view
  - [ ] How will 3D models look in our viewer? Will they have show their default textures, show a single color, etc.? This isn't an CC-licensed asset, but here's [an example](https://creazilla.com/nodes/3061-lion-3d-model) of viewing the original textures *and* a plain color model.
  - [ ] What will the interaction look like? There should be a 'play' button, i.e, models aren't loaded by default. Some sites, like [thingiverse](https://www.thingiverse.com/thing:5238596), include actual photographs of the models, not just the 3d renders. How/where do we display those? Like album art, with the metadata, or some other way?
  - [ ] Any special metadata UI we need that doesn't already exist?

**Comment by sarayourfriend** [comment]
> How will 3D models look in our viewer? Will they have show their default textures, show a single color, etc.? This isn't an CC-licensed asset, but here's [an example](https://creazilla.com/nodes/3061-lion-3d-model) of viewing the original textures and a plain color model.

It didn't occur to me that for MVP for this we could avoid having to solve the interactive viewer problem altogether and just choose some angled poses and static renders like the link you shared does. It would also let us defer figuring out how to accommodate low power devices.

I took a closer look at Trois a couple weekends ago and it needs a _lot_ of work to make it Vue 2/3 compatible. It makes heavy usage of the Options API in ways that would be difficult and take a long time to help pick apart into composables. We could do it, but it would probably take a lot of someone's time on the team over a long period of time. And it seems like the maintainer is (understandably) more interested in maintaining and enhancing existing features rather than adding Vue 2 support.

File: rfcs/20220221-3d_model_support.md
Comment (sarayourfriend) [inline rfcs/20220221-3d_model_support.md:3]: Tentatively approving. The backend plan seems good to me. I think we'll want another RFC with more specific implementation details once we have frontend designs but this has been a really good starting point for having us discuss and think about the various hurdles we'll need to make sure designs are aware of.

```suggestion
- [z] @sarayourfriend
```

**Review by sarayourfriend — APPROVED** [review]

File: rfcs/20220221-3d_model_support.md
Comment (AetherUnbound) [reply rfcs/20220221-3d_model_support.md:36]: I love it, thanks @panchovm!

**Review by panchovm — APPROVED** [review]

File: rfcs/20220221-3d_model_support.md
Comment (zackkrida) [inline rfcs/20220221-3d_model_support.md:4]:
```suggestion
- [x] @panchovm 
```

**Pull request merged**
